### PR TITLE
Make saving and loading the responsibility of Tox rather than Messenger

### PR DESCRIPTION
**Pull request by zugz**
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1213)

### CODE REVIEW

File: auto_tests/save_load_test.c [hunk 1]
```diff
@@ -88,18 +88,43 @@ static void test_few_clients(void)
...
+     * b) a saved state can be loaded back successfully
+     * c) a second save is of equal size
+     * d) the second save is of equal content */
+    const size_t extra = 64;
     const size_t save_size1 = tox_get_savedata_size(tox2);
     ck_assert_msg(save_size1 != 0, "save is invalid size %u", (unsigned)save_size1);
     printf("%u\n", (unsigned)save_size1);
-    VLA(uint8_t, save1, save_size1);
-    tox_get_savedata(tox2, save1);
+    VLA(uint8_t, buffer, save_size1 + 2 * extra);
```
Comment (iphydf) [inline auto_tests/save_load_test.c]: Perhaps while you're changing this code already, remove the vla and use malloc/free?

File: toxcore/group.h [hunk 2]
```diff
@@ -448,6 +448,20 @@ int callback_groupchat_peer_delete(Group_Chats *g_c, uint32_t groupnumber, peer_
  */
 int callback_groupchat_delete(Group_Chats *g_c, uint32_t groupnumber, group_on_delete_cb *function);
 
+/* Return size of the conferences data (for saving). */
+uint32_t conferences_size(const Group_Chats *g_c);
+
+/* Save the conferences in data (must be allocated memory of size at least conferences_size()) */
+uint8_t *conferences_save(const Group_Chats *g_c, uint8_t *data);
+
+/* Load a section.
```
Comment (iphydf) [inline toxcore/group.h]:
```c
/**
 * Load a section.
 *
 * @...
```

File: toxcore/state.c [hunk 3]
```diff
@@ -103,3 +103,21 @@ void lendian_to_host32(uint32_t *dest, const uint8_t *lendian)
 #endif
     *dest = d;
 }
+
+void host_to_lendian_bytes16(uint8_t *dest, uint16_t num)
+{
+#ifdef WORDS_BIGENDIAN
```
Comment (iphydf) [inline toxcore/state.c]: At some point we should stop depending on these and write portable code. Do you want to do that now or defer it?

Comment (zugz) [reply toxcore/state.c]: Hmm... I wouldn't know how best to do it, and I understand we intend to redo saving completely eventually. So I'd rather defer.

**Comment by zugz** [comment]
I factored the save invariance and associated tests out as a separate 
PR, #1215.

File: toxcore/tox.c [hunk 4]
```diff
@@ -328,6 +328,51 @@ bool tox_version_is_compatible(uint32_t major, uint32_t minor, uint32_t patch)
...
+    uint32_t cookie_len = sizeof(data32);
+
+    if (length < cookie_len) {
+        return -1;
+    }
+
+    memcpy(data32, data, sizeof(uint32_t));
+    lendian_bytes_to_host32(data32 + 1, data + sizeof(uint32_t));
+
+    if (!data32[0] && (data32[1] == STATE_COOKIE_GLOBAL)) {
```
Comment (iphydf) [inline toxcore/tox.c]: Replace `!num` with `num == 0` when `num` is an integer, not a bool. Also, would it make sense to return -1 on the inverted condition, considering it an error return, and returning the success/failure of state_load at the end of the function?

File: toxcore/tox.c [hunk 5]
```diff
@@ -537,17 +582,42 @@ void tox_kill(Tox *tox)
...
     const Messenger *m = tox->m;
-    return messenger_size(m);
+    return 2 * sizeof(uint32_t)
+           + messenger_size(m)
+           + conferences_size(m->conferences_object)
+           + end_size();
 }
 
-void tox_get_savedata(const Tox *tox, uint8_t *savedata)
+void tox_get_savedata(const Tox *tox, uint8_t *data)
```
Comment (iphydf) [inline toxcore/tox.c]: Why did you rename savedata to data? Now the .c and .h files are out of sync.

**Comment by zugz** [comment]
* Tuesday, 2018-10-16 at 15:23 -0700 - iphydf <notifications@github.com>:

>Can you rebase or merge upstream/master?

Done

**Comment by zugz** [comment]
* Wednesday, 2018-10-17 at 05:17 -0700 - iphydf <notifications@github.com>:

>*[testing/Messenger_test.c, line 137 at r3](https://reviewable.io/reviews/toktok/c-toxcore/1213#-LP0w4_o8qCfdpnPxHUy:-LP0w4_o8qCfdpnPxHUz:bqszos9) ([raw file](https://github.com/toktok/c-toxcore/blob/7f90dca13d91a2e47f859f31a29dc12056e7f40a/testing/Messenger_test.c#L137)):* 
>> ```C 
>>     }
>>
>>     /* Loading disabled, because loading is now handled by Tox rather than Messenger */ 
>> ```
>
>Should this also be deleted instead of commented out? 

I'd rather leave it, in case someone one day wants to adapt it.


Done the rest. I can fire up reviewables to say so there if that would 
really be helpful.
